Priority in Microgrid Controller Selection

iguessthislldo opened the pull request:
Fixes https://github.com/OpenDDS/DeveloperResources/issues/12

iguessthislldo committed:
Priority in Microgrid Controller Selection

Fixes https://github.com/OpenDDS/DeveloperResources/issues/12

iguessthislldo committed:
Merge branch 'main' into igtd/mc-sel-priority

iguessthislldo committed:
Add Testing to GHA and Tweaks

mitza-oci reviewed (changes requested):
Also see warnings in ControllerSelector.cpp, and CI errors

iguessthislldo commented on tactical-microgrid-standard/common/ControllerSelector.h:
```diff
@@ -109,6 +145,30 @@ class OpenDDS_TMS_Export ControllerSelector :
+    uint16_t priority = 0;
+    tms::Identity id;
+
+    explicit PrioritizedController(const tms::DeviceInfo& di)
+    : priority(0)
+    , id(di.deviceId())
```
Not sure why I did that after defaulting the member.

iguessthislldo committed:
Respond to Review and Use SIGINT

iguessthislldo committed:
Remove Testing for macOS

macOS seems to have issues with IPv6 Multicast.

iguessthislldo committed:
Workaround Escaping in opnedds_add_test

https://github.com/OpenDDS/OpenDDS/pull/5071

iguessthislldo committed:
Use qw

iguessthislldo committed:
Move Windows Stack Change to TMS_Common

This allows all programs to use it.

iguessthislldo committed:
Don't do SIGINT on Windows and try trace on Linux

iguessthislldo committed:
Add Missing OS if

iguessthislldo committed:
Try Kill(..., 0) on Windows

iguessthislldo committed:
Explicit Shutdown in Test

iguessthislldo committed:
More Fixes For Occasional Failures

- Logging for heartbeats.
- Give more time for mc3 to be selected.
- Don't base mc kill timeout on when it was supposed to stop.
- Use double-based duration to avoid situations when time duration math
  results are rounded to zero.

iguessthislldo committed:
Try Removing IPv6 and Log Ignore

iguessthislldo committed:
Put --ipv6 Back, Try TARGET_RUNTIME_DLL_DIRS

iguessthislldo committed:
Try mc-sel Test on macOS Without IPv6

iguessthislldo committed:
Revert "Try mc-sel Test on macOS Without IPv6"

This reverts commit 0f536596eee308bff057b9fea848d41ad68e2f3e.

The mc-sel Test can work in GHA, but the controllers seem to have a high
chance of crashing when the run_test.pl sends a SIGINT.

sonndinh commented on tactical-microgrid-standard/common/ControllerSelector.cpp:
```diff
@@ -14,6 +18,9 @@ void ControllerSelector::got_heartbeat(const tms::Heartbeat& hb)
   Guard g(lock_);
   auto it = all_controllers_.find(hb.deviceId());
   if (it != all_controllers_.end()) {
+    ACE_DEBUG((LM_DEBUG, "(%P|%t) ControllerSelector::got_heartbeat: from mc \"%C\"\n",
+      hb.deviceId().c_str()));
```
Could this and similar log messages be conditional, since they could produce a lot of logs in a live run? An option is a `--verbose` command-line option from the concrete devices that can be passed down here to use.

iguessthislldo replied:
Yes, I can. I committed them because I had to keep using them to figure out what was alive at a glance both locally and in CI.

sonndinh commented on tactical-microgrid-standard/controller/Controller.h:
```diff
@@ -6,7 +6,11 @@
 class Controller : public Handshaking {
 public:
-  explicit Controller(const tms::Identity& id) : Handshaking(id) {}
+  explicit Controller(const tms::Identity& id, uint16_t priority = 0)
+  : Handshaking(id)
+  , priority_(priority)
```
```suggestion
  explicit Controller(const tms::Identity& id, uint16_t priority = 0)
    : Handshaking(id)
    , priority_(priority)
```

sonndinh commented on tactical-microgrid-standard/power_devices/PowerDevice.h:
```diff
@@ -87,6 +92,11 @@ class PowerSim_Idl_Export PowerDevice : public Handshaking {
     return ret;
   }
 
+  void delete_extra_entities()
+  {
+    delete_entities(sim_participant_);
+  }
```
Nice catch!

iguessthislldo committed:
Configurable Debug Messages

iguessthislldo committed:
Remove Commented Line
